Work around OTP-21 deprecation of get_stacktrace() and other incompatible changes

ferd opened the pull request:
Depends on https://github.com/erlware/erlware_commons/pull/130 to succeed

Fixes https://github.com/erlang/rebar3/issues/1772

ferd commented:
To discuss: should we just remove the stacktraces where we can afford it rather than duplicating all of that code?

tsloughter commented:
I published erlware common 1.1.0.

ferd commented:
So the change in rebar3.erl actually breaks the entire stacktrace/crashdump reporting for uncaught errors. So that needs more work.

tsloughter commented:
Well shit. I thought we must translate an exception to an error tuple somewhere that we could add a match on stacktrace and add it to the tuple, but not seeing where that was done. 

Will look more tomorrow.

tsloughter commented:
Oh, you got it, nevermind :)

ferd commented:
Yeah. Right now I'm checking stuff and I blew up Dialyzer by probably fat-fingering something. I need some better QA on this.

ferd commented:
Alright, I fixed Dialyzer, but rebar3 shell is now tricky because the `error_logger:swap_handler(tty)` function has been removed with the new logger API, and from what @sirihansen told me today, there is no direct equivalent of it anymore. 

We would have a regression on https://github.com/erlang/rebar3/pull/1158

ferd commented:
I manually confirmed that the new logger has no hangup with the group leaders of before, so I rewrote the thing to avoid some problems. Dialyzer will keep complaining about it but for now that seems good enough.

ferd commented:
So, OTP 20 had a warning preventing the usage of get_stacktrace() outside of a try .. catch expression. Not sure why it ever passed before, but now the code no longer compiles only on that version in rebar3.erl. Will need to dig some more.

ferd commented:
Reworked the pull request based off the macros by @okeuday at https://github.com/erlang/otp/pull/1783#issuecomment-386190970 -- I also reordered and squashed commits so they work better in the right order.
